### PR TITLE
Fix 2 DMDc ROM issues

### LINKED ISSUE #1932: [DEFECT] DMDc ROM with multiple inputs
### Thank you for the defect report

- [X] I am using the latest version of `RAVEN`.
- [X] I have read the [Wiki](https://github.com/idaholab/raven/wiki).
- [X] I have created a [minimum, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)
that demonstrates the defect.

### Defect Description

[Description]

If one tries to construct a DMDc ROM with multiple inputs, RAVEN seems constructing a ROM, but it fails to run ROM (i.e., generate samples by using the ROM) and returns an error.

< Error Message >
ErrorMessage_1

### Steps to Reproduce

[Steps to Reproduce]

1) Open the csv file:
\..\raven\tests\framework\ROM\TimeSeries\DMD\DMDC\trainingData\Synthesized_BOP_0.csv

2) Add a column called u2. To avoid collinearity between u1 and u2, let us create u2 using the following formula in excel sheet 

< Screen Shot of Input Data File >
screen_shot_CSV_input

3) Open RAVEN input file
\..\raven\tests\framework\ROM\TimeSeries\DMD\test_unparameterized_dmdc.xml

4) Add u2 in relevant nodes. u2-related code lines need to be placed in "Models", "Samplers", and "DataObjects" nodes. Please be aware that I uploaded the RAVEN input file I used. 

< RAVEN Code snippet >
RAVEN_code_snippet

5) When one runs the RAVEN input deck, the following error message comes out. 

< Error Message >
ErrorMessage_3

### Expected Behavior

RAVEN should generate samples using the DMDc ROM. 

### Screenshots and Input Files

I uploaded one zip file containing:
1) BOP_Data_index_unPara.csv (TrainDataFile in <Files> node of RAVEN input deck)
2) Synthesized_BOP_0.csv (data input file containing u2 as well)
3) test_unparameterized_dmdc.xml (RAVEN input deck I used)

[test_unparameterized_dmdc.zip](https://github.com/idaholab/raven/files/9283188/test_unparameterized_dmdc.zip)

### OS

Windows

### OS Version

_No response_

### Dependency Manager

CONDA

### For Change Control Board: Issue Review

- [x] Is it tagged with a type: defect or task?
- [x] Is it tagged with a priority: critical, normal or minor?
- [x] If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

### For Change Control Board: Issue Closure

- [x] If the issue is a defect, is the defect fixed?
- [x] If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1934: [DEFECT] DMDc ROM unpickling isssue
### Thank you for the defect report

- [X] I am using the latest version of `RAVEN`.
- [X] I have read the [Wiki](https://github.com/idaholab/raven/wiki).
- [X] I have created a [minimum, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)
that demonstrates the defect.

### Defect Description

DMDc ROM created in RAVEN environment cannot be unpickled by ravenROMexternal class in externalROMloader.py (raven/scripts/externalROMloader.py).

The figure below is the error message I have got when I tried to unpickle it. 

<Error Message>
DMDc_ROM_unpickling_issue

### Steps to Reproduce

1)	Open RAVEN input deck:
\...\projects\raven\tests\framework\ROM\TimeSeries\DMD\test_unparameterized_dmdc.xml

2)	Change some nodes (e.g. <RunInfo>, <Files>, <Steps>) so that the ROM can be dumped (I attached the input deck I used).

3)	Run the Python code attached (You need to change the path accordingly). This Python code was developed to unpickle the file using the ravenROMexternal class in externalROMloader.py. 

### Expected Behavior

RAVEN should be able to unpickle the pickled file. 

### Screenshots and Input Files

In the attached zip file, there are two files:

1) junyung_test_unparameterized_dmdc.xml - RAVEN input deck used to create a picked file using DMDc ROM.
2) Junyung_LoadingRAVENpickledRomsExternaly.ipynb - Jupyter notebook used for testing unpickling. 

[Junyung_LoadingRAVENpickledRomsExternaly.zip](https://github.com/idaholab/raven/files/9284773/Junyung_LoadingRAVENpickledRomsExternaly.zip)

### OS

Windows

### OS Version

_No response_

### Dependency Manager

CONDA

### For Change Control Board: Issue Review

- [x] Is it tagged with a type: defect or task?
- [x] Is it tagged with a priority: critical, normal or minor?
- [x] If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

### For Change Control Board: Issue Closure

- [x] If the issue is a defect, is the defect fixed?
- [x] If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by dgarrett622**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1934 
Closes #1932 

##### What are the significant changes in functionality due to this change request?
DMDc ROMs were unable to be unpickled with externalROMloader because `__setstate__` method included self.featureVals which goes unused in DMDc ROM. This fix enables unpickling DMDc ROMs with externalROMloader.

DMDc ROMs would throw an error when run if more than one input/actuator was included due to two array broadcasting and shape issues. This fix corrects the issues and allows DMDc ROM to have more than one input/actuator.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by dgarrett622** [comment]
Since it also involves the DMDc ROM, I'm going to include a fix for #1932 in this PR as well.

File: ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py [hunk 1]
```diff
@@ -245,19 +256,21 @@ def __evaluateLocal__(self,featureVals):
...
+        # use np.expand_dims to ensure this works with multiple actuators
+        uVector = uVector - np.expand_dims(self.actuatorVals[0, index, :], axis=tuple(range(1,len(uVector.shape))))
         initStates = initStates - self.stateVals[0, index, :]
       evalX[cnt, 0, :] = initStates
       evalY[cnt, 0, :] = np.dot(self.__Ctilde[index, :, :], evalX[cnt, 0, :])
       ### perform the self-propagation of X, X[k+1] = A*X[k] + B*U[k] ###
       for i in range(tsEval-1):
-        xPred = np.reshape(self.__Atilde[index,:,:].dot(evalX[cnt,i,:]) + self.__Btilde[index,:,:].dot(uVector[cnt,:,i]),(-1,1)).T
+        # make sure that Btilde dot uVector works correctly for multiple inputs and has the same shape as aTilde dot evalX
+        xPred = np.reshape(self.__Atilde[index,:,:].dot(evalX[cnt,i,:]) + (self.__Btilde[index,:,:].dot(uVector[:,:,i])).reshape(-1,), (-1,1)).T
```
Comment (wangcj05) [inline ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py:273]: I'm not sure why 'cnt' index is not present in uVector anymore. I would like @wanghy-anl  to take a look at this PR.

**Review by wangcj05 — COMMENTED** [review]
The code changes are good to me. There are something that I could not verify. @wanghy-anl Could you take a look at this PR? 
In addition, @dgarrett622 Is it possible to add a test for this PR?

File: ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py (hunk 1 above)
Comment (dgarrett622) [reply ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py:273]: Including the 'cnt' index in uVector results in cycling through each actuator individually, B*U for B (n x m) and U (m x p) results in B(n x m)*U(1 x p). For only one input u, there is no error, but as soon as you add another input the error pops up.

**Comment by dgarrett622** [comment]
> The code changes are good to me. There are something that I could not verify. @wanghy-anl Could you take a look at this PR? In addition, @dgarrett622 Is it possible to add a test for this PR?

I can put together a test based on Junyung's inputs.

File: ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py (hunk 1 above)
Comment (wanghy-anl) [reply ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py:273]: > Including the 'cnt' index in uVector results in cycling through each actuator individually, B*U for B (n x m) and U (m x p) results in B(n x m)*U(1 x p). For only one input u, there is no error, but as soon as you add another input the error pops up.

I have some reservations here. In alfoa/feature_selection branch we implemented it differently. Let me do some test on my side and get back to you.

Comment (wanghy-anl) [reply ravenframework/SupervisedLearning/DynamicModeDecompositionControl.py:273]: @dgarrett622 @wangcj05 This removal of `cnt` in `uVector` is OK. `cnt` always have a value of zero during the evaluation step, and like Daniel explained, we need to do the dot product for all the m variables in `uVector[:,:,i]`. I tested the newly added line #273 in alfoa/feature_selection branch and it returned the same result.

**Review by wangcj05 — APPROVED** [review]
PR looks good.

**Comment by wangcj05** [comment]
checklist is good. Tests are green.

**Pull request merged**